course_UvA-DL/activation-functions/Activation_Functions.py

```diff
@@ -0,0 +1,824 @@
+    # we set this variable to 0. After running through the whole training set, only dead neurons will have a 1.
+    neurons_dead = [
+        torch.ones(layer.weight.shape[0], device=device, dtype=torch.bool) for layer in net.layers[:-1]
+        if isinstance(layer, nn.Linear)
+    ]  # Same shapes as hidden size in BaseNetwork
+
+    net.eval()
+    with torch.no_grad():
+        for imgs, labels in tqdm(train_loader, leave=False):  # Run through whole training set
+            layer_index = 0
+            imgs = imgs.to(device)
+            imgs = imgs.view(imgs.size(0), -1)
+            for layer in net.layers[:-1]:
+                imgs = layer(imgs)
+                if isinstance(layer, ActivationFunction):
+                    # Are all activations == 0 in the batch, and we did not record the opposite in the last batches?
+                    neurons_dead[layer_index] = torch.logical_and(
+                        neurons_dead[layer_index], (imgs == 0).all(dim=0)
+                    )
+                    layer_index += 1
+    number_neurons_dead = [t.sum().item() for t in neurons_dead]
+    print("Number of dead neurons:", number_neurons_dead)
+    print(
+        "In percentage:", ", ".join([
+            f"{(100.0 * num_dead / tens.shape[0]):4.2f}%"
+            for tens, num_dead in zip(neurons_dead, number_neurons_dead)
+        ])
+    )
+
+
+# %% [markdown]
+# First, we can measure the number of dead neurons for an untrained network:
+
+# %%
+set_seed(42)
+net_relu = BaseNetwork(act_fn=ReLU()).to(device)
+measure_number_dead_neurons(net_relu)
+
+# %% [markdown]
+# We see that only a minor amount of neurons are dead, but that they increase with the depth of the layer.
+# However, this is not a problem for the small number of dead neurons we have as the input to later layers is changed due to updates to the weights of previous layers.
+# Therefore, dead neurons in later layers can potentially become "alive"/active again.
+#
+# How does this look like for a trained network (with the same initialization)?
+
+# %%
+net_relu = load_model(model_path=CHECKPOINT_PATH, model_name="FashionMNIST_relu").to(device)
+measure_number_dead_neurons(net_relu)
+
+# %% [markdown]
+# The number of dead neurons indeed decreased in the later layers.
+# However, it should be noted that dead neurons are especially problematic in the input layer.
+# As the input does not change over epochs (the training set is kept as it is), training the network cannot turn those neurons back active.
+# Still, the input data has usually a sufficiently high standard deviation to reduce the risk of dead neurons.
+#
+# Finally, we check how the number of dead neurons behaves with increasing layer depth.
+# For instance, let's take the following 10-layer neural network:
+
+# %%
+set_seed(42)
+net_relu = BaseNetwork(act_fn=ReLU(), hidden_sizes=[256, 256, 256, 256, 256, 128, 128, 128, 128,
+                                                    128]).to(device)
```

Comment:
```suggestion
                                                    128],).to(device)
```
trailing comma shall make this as multilne